JDL fails to compile: bad type in conditional expression

##### **Overview of the issue**

I'm trying to re-create the backend I used in my [Mobile JHipster](https://developer.okta.com/blog/2020/04/27/mobile-development-ionic-react-native-jhipster) presentation. The app I'm trying to create is a bit like my 21-Points Health app, but I changed it to use OAuth, React, and WebFlux.

```
/*
 * 21-Points (https://github.com/mraible/21-points) as a WebFlux + React app
 */

application {
  config {
    applicationType monolith
    reactive true
    baseName HealthPoints
    packageName com.okta.developer
    authenticationType oauth2
    prodDatabaseType postgresql
    buildTool gradle
    testFrameworks [cypress]
    clientFramework react
    enableTranslation true
    nativeLanguage en
    languages [en, es]
  }
  entities Points, BloodPressure, Weight, Preferences
}

// JDL definition for application 'TwentyOnePoints' generated with command 'jhipster export-jdl'

entity BloodPressure {
  timestamp ZonedDateTime required,
  systolic Integer required,
  diastolic Integer required
}
entity Weight {
  timestamp ZonedDateTime required,
  weight Double required
}
entity Points {
  date LocalDate required,
  exercise Integer,
  meals Integer,
  alcohol Integer,
  notes String maxlength(140)
}
entity Preferences {
  weeklyGoal Integer required min(10) max(21),
  weightUnits Units required
}

enum Units {
  KG,
  LB
}

relationship OneToOne {
  Preferences{user(login)} to User
}
relationship ManyToOne {
  BloodPressure{user(login)} to User,
  Weight{user(login)} to User,
  Points{user(login)} to User
}

paginate BloodPressure, Weight with infinite-scroll
paginate Points with pagination
```

Running `jhipster jdl app.jdl` works fine, but when I try to compile it, I get an error:

```
> Task :compileJava
Note: Hibernate JPA 2 Static-Metamodel Generator 5.4.28.Final
/Users/mraible/dev/mobile-jhipster/backend/src/main/java/com/okta/developer/domain/Preferences.java:83: error: incompatible types: bad type in conditional expression
        this.userId = user != null ? user.getId() : null;
                                               ^
    String cannot be converted to Long
/Users/mraible/dev/mobile-jhipster/backend/src/main/java/com/okta/developer/domain/Preferences.java:89: error: incompatible types: bad type in conditional expression
        this.userId = user != null ? user.getId() : null;
                                               ^
    String cannot be converted to Long
/Users/mraible/dev/mobile-jhipster/backend/src/main/java/com/okta/developer/domain/Preferences.java:93: error: incompatible types: Long cannot be converted to String
        return this.userId;
                   ^
/Users/mraible/dev/mobile-jhipster/backend/src/main/java/com/okta/developer/domain/Preferences.java:97: error: incompatible types: String cannot be converted to Long
        this.userId = user;
                      ^
Note: /Users/mraible/dev/mobile-jhipster/backend/src/main/java/com/okta/developer/service/ColumnConverter.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
4 errors

> Task :compileJava FAILED

FAILURE: Build failed with an exception.
```

##### **Motivation for or Use Case**

It would be nice if JDLs I previously used still worked.

##### **Reproduce the error**

Copy the JDL above and run `jhipster jdl` to import it. Then run `./gradlew`.

##### **JHipster Version(s)**

`main` branch as of March 15, 2021, 18:00 MDT

--- a/generators/entity-server/templates/src/main/java/package/domain/Entity.java.ejs
+++ b/generators/entity-server/templates/src/main/java/package/domain/Entity.java.ejs
@@ -475,7 +475,7 @@ relationships.forEach((relationship, idx) => {
         <%_ }
         } else if (databaseType === 'sql' && reactive) { _%>
           <%_ if (ownerSide === true && isUsingMapsId !== true) { _%>
-    private Long <%= relationshipFieldName %>Id;
+    private <%= relationship.otherEntity.primaryKey.type %> <%= relationshipFieldName %>Id;
 
           <%_ } _%>
     @Transient

--- a/generators/server/templates/src/main/java/package/service/ColumnConverter.java.ejs
+++ b/generators/server/templates/src/main/java/package/service/ColumnConverter.java.ejs
@@ -21,7 +21,6 @@ package <%= packageName %>.service;
 import org.springframework.core.convert.ConversionService;
 import org.springframework.data.r2dbc.convert.R2dbcConverter;
 import org.springframework.data.r2dbc.convert.R2dbcCustomConversions;
-import org.springframework.data.r2dbc.core.ReactiveDataAccessStrategy;
 import org.springframework.lang.Nullable;
 import org.springframework.stereotype.Service;
 import org.springframework.util.ClassUtils;
@@ -49,6 +48,7 @@ public class ColumnConverter {
      * @param <T> the parameter for the intended type.
      * @return the value which can be constructed from the input.
      */
+    @SuppressWarnings("unchecked")
     public <T> T convert(@Nullable Object value, @Nullable Class<T> target) {
 
         if (value == null || target == null || ClassUtils.isAssignableValue(target, value)) {

